File: src/main/java/net/sf/rails/game/model/CalculatedMoneyModel.java
package net.sf.rails.game.model;

import net.sf.rails.game.RailsItem;
import net.sf.rails.game.state.Currency;

/**
 * This is MoneyModel that derives it value from a calculation method.
 * TODO: Rewrite all methods implementing the interface
 */
public final class CalculatedMoneyModel extends MoneyModel {

    public interface CalculationMethod {
        public int calculate();
        public boolean initialised();
    }
    
    private final CalculationMethod method;

    private CalculatedMoneyModel(RailsItem parent, String id, CalculationMethod method, Currency currency) {
        super(parent, id, currency);
        this.method = method;
    }

    public static CalculatedMoneyModel create(RailsItem parent, String id, CalculationMethod method) {
        Currency currency = parent.getRoot().getBank().getCurrency();
        return new CalculatedMoneyModel(parent, id, method, currency);
    }
    
    @Override
    public int value() {
        return method.calculate();
    }
   
    @Override
    public boolean initialised() {
        return method.initialised();
    }

}

File: dotnet/RailsLib.Net/Net/Game/Model/CalculatedMoneyModel.cs
using GameLib.Net.Game.State;
using System;
using System.Collections.Generic;
using System.Text;

/**
 * This is MoneyModel that derives it value from a calculation method.
 * TODO: Rewrite all methods implementing the interface
 */

namespace GameLib.Net.Game.Model
{
    public interface ICalculationMethod
    {
        int Calculate();
        bool Initialized { get; }
    }

    abstract public class CalculationMethodBase : ICalculationMethod
    {
        abstract public int Calculate();
        abstract public bool Initialized { get; }
    }

    public class CalculatedMoneyModel : MoneyModel
    {

        private ICalculationMethod method;

        private CalculatedMoneyModel(IRailsItem parent, string id, ICalculationMethod method, Currency currency) :
                base(parent, id, currency)
        {
            this.method = method;
        }

        public static CalculatedMoneyModel Create(IRailsItem parent, string id, ICalculationMethod method)
        {
            Currency currency = parent.GetRoot.Bank.Currency;
            return new CalculatedMoneyModel(parent, id, method, currency);
        }

        override public int Value
        {
            get
            {
                return method.Calculate();
            }
        }

        override public bool Initialized
        {
            get
            {
                return method.Initialized;
            }
        }
    }
}
